lib/puppet/provider/foreman_hostgroup/rest_v3.rb

```diff
@@ -0,0 +1,186 @@
+      r = request(:get, path, search: %(title="#{search_title}" and name="#{search_name}"))
+
+      raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}" unless success?(r)
+
+      results = JSON.parse(r.body)['results']
+      unless results.empty?
+        raise Puppet::Error, "Too many hostgroups found when looking for hostgroup with name #{search_name} and title #{search_title}" if results.size > 1
+
+        get_hostgroup_by_id(results[0]['id'])
+      end
+    end
+  end
+
+  def get_hostgroup_by_id(id)
+    path = "api/v2/hostgroups/#{id}"
+    r = request(:get, path)
+
+    return JSON.parse(r.body) if success?(r)
+
+    raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+  end
+
+  def id
+    hostgroup ? hostgroup['id'] : nil
+  end
+
+  def exists?
+    !id.nil?
+  end
+
+  def parent_hostgroup
+    return nil unless resource[:parent_hostgroup]
+
+    @parent_hostgroup ||= begin
+      path = 'api/v2/hostgroups'
+      search_title = resource[:parent_hostgroup]
+      search_name = if resource[:parent_hostgroup_name]
+                      resource[:parent_hostgroup_name]
+                    elsif %r{^(?<_parent>.+)/(?<name>.+)$} =~ resource[:parent_hostgroup]
+                      name
+                    else
+                      search_title
+                    end
+      Puppet.debug("Searching for parent hostgroup with name #{search_name} and title #{search_title}")
+      r = request(:get, path, search: %(title="#{search_title}" and name="#{search_name}"))
+
+      raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}" unless success?(r)
+
+      results = JSON.parse(r.body)['results']
+      raise Puppet::Error, "Parent hostgroup #{resource[:parent_hostgroup]} for #{resource[:name]} not found" if results.empty?
+      raise Puppet::Error, "Too many hostgroups found when looking for parent hostgroup with name #{search_name} and title #{search_title}" if results.size > 1
+
+      get_hostgroup_by_id(results[0]['id'])
+    end
+  end
+
+  def parent_hostgroup_id
+    parent_hostgroup ? parent_hostgroup['id'] : nil
+  end
+
+  def organization_id(organization_title)
+    title_to_id('organizations', organization_title)
+  end
+
+  def location_id(location_title)
+    title_to_id('locations', location_title)
+  end
+
+  # Returns the id of a location/organization based on its title
+  def title_to_id(type, title)
+    path = "api/v2/#{type}"
+    r = request(:get, path, search: %(title="#{title}"))
+    unless success?(r)
+      raise Puppet::Error,
+            "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+    end
+    results = JSON.parse(r.body)['results']
+    raise Puppet::Error, "#{title} not found in #{type}" unless results.size == 1
+
+    results[0]['id']
+  end
```

Comment:
Not that important, but I wonder if these could be in some `private` part to indicate what's public API and what are helpers.